XMonad/Layout/IndependentScreens.hs

```diff
@@ -106,8 +111,63 @@ withScreens :: ScreenId            -- ^ The number of screens to make workspaces
             -> [PhysicalWorkspace] -- ^ A list of all internal physical workspace names
 withScreens n vws = [marshall sc pws | pws <- vws, sc <- [0..n-1]]
 
-onCurrentScreen :: (VirtualWorkspace -> WindowSet -> a) -> (PhysicalWorkspace -> WindowSet -> a)
-onCurrentScreen f vws = screen . current >>= f . flip marshall vws
+-- | Transform a function over physical workspaces into a function over virtual workspaces.
+-- This is useful as it allows you to write code without caring about the current screen, i.e. to say "switch to workspace 3"
+-- rather than saying "switch to workspace 3 on monitor 3".
+onCurrentScreen :: (PhysicalWorkspace -> WindowSet -> a) -> (VirtualWorkspace -> WindowSet -> a)
+onCurrentScreen f vws ws =
+  let currentScreenId = W.screen $ W.current ws
+   in f (marshall currentScreenId vws) ws
+
+-- | Get the workspace currently active on a given screen
+workspaceOnScreen :: ScreenId -> WindowSet -> Maybe PhysicalWorkspace
+workspaceOnScreen screenId ws = W.tag . W.workspace <$> screenOnMonitor screenId ws
+
+
+-- | generate WindowSet transformation by providing a given function with the workspace active on a given screen.
+-- This may for example be used to shift a window to another screen as follows:
+--
+-- > windows $ withWspOnScreen 1 W.shift
+--
+withWspOnScreen :: ScreenId                               -- ^ The screen to run on
+         -> (PhysicalWorkspace -> WindowSet -> WindowSet) -- ^ The transformation that will be passed the workspace currently active on there
+         -> WindowSet -> WindowSet
+withWspOnScreen screenId operation ws = case workspaceOnScreen screenId ws of
+    Just wsp -> operation wsp ws
+    Nothing -> ws
+
+-- | Get the workspace that is active on a given screen.
+screenOnMonitor :: ScreenId -> WindowSet -> Maybe (W.Screen WorkspaceId (Layout Window) Window ScreenId ScreenDetail)
+screenOnMonitor screenId ws = find ((screenId ==) . W.screen) (W.current ws : W.visible ws)
+
+-- | Focus a window, switching workspace on the correct Xinerama screen if neccessary.
+focusWindow' :: Window -> WindowSet -> WindowSet
+focusWindow' window ws
+  | Just window == W.peek ws = ws
+  | otherwise = case W.findTag window ws of
+      Just tag -> W.focusWindow window $ focusScreen (unmarshallS tag) ws
+      Nothing -> ws
+
+-- | Focus a given screen.
+focusScreen :: ScreenId -> WindowSet -> WindowSet
+focusScreen screenId = withWspOnScreen screenId W.view
+
+-- | Get the nth virtual workspace
+nthWorkspace :: Int -> X (Maybe VirtualWorkspace)
+nthWorkspace n = (!!? n) . workspaces' <$> asks config
+
+
+-- | HandleEventHook which makes the activate window event respect IndependentScreens.
+-- Without this, a window requesting activation may cause a workspace switch
+-- that shows a workspace from a different screen on the current one.
+activateWindowEventHook :: Event -> X All
+activateWindowEventHook ClientMessageEvent { ev_message_type = messageType, ev_window = window } = do
```

Comment:
Having a custom event hook here that conflicts with what the EwmhDesktops event hook does seems wrong. On the other hand, until #399 is finished, trying to do this correctly might be a waste of time. :-/

(Also, a documentation explaining how this should be used and what the caveats are is missing.)

Reply:
Do you see any better way to implement this right now, or should i - given that #399 exists, just remove this part?

Reply:
You can make this a ManageHook so that it works with what was merged in #192 but any documentation that you write for it will need to be adapted after #399.